### ISSUE TITLE
Allow multiple operations to be executed in one call

### ISSUE THREAD

**Issue by bbakerman** [graphql-java/graphql-java#806]
Today you can only execute 1 operation.  The spec says you can only do that but I feel its an unnecessary restriction.

I would like graphql-java to introduce a experimental mode where you can provide a list of operation names to execute

So given this valid query

```
query Orders {
      orders {
        id
        items {
          name
       }
    }
}

query Billing {
    billing {
        dueDate
        amount
    }
}
```

We could execute both operations by specify the operation names in the ExecutionInput

```
    ExecutionInput input = newInput().operations(toList("A","B")).build();
```

We would retain the current single operation name for backwards compatiblity.

If you did this then you need to get a combined execution result.  The names of the operations would be the top level values

```
   executionResult == {
          "A" : {
              "orders" : [ ...]
          },
         "B" : {
              "billing" : {
                 ...
              }
         }
   }
```

The errors would be a combined list of executing operation A and then operation B

The operations would be executed in the order specified in the input list

Looking at the scope of changes (and spec enhancement) I think this is a good candidate for some library extension.

This is somewhat inspired from the Sangria implementation here : http://sangria-graphql.org/learn/#batch-executor

**Comment by bbakerman** [comment]
This is an example test that might pass with this mechanism

        GraphQL graphQL = newGraphQL(starWarsSchema).build()

        def query = '''
            query A {
                hero {
                    name
                }
            }    

            query B {
                human(id:"1000") {
                    name
                }
            }    
        '''
        ExecutionInput input = newExecutionInput().operationNames(["A","B"]).query(query).build()

        def er = graphQL.execute(input)

        then:
        er.errors.size() == 0
        er.data["A"] == [hero: [name: "R2-D2"]]
        er.data["B"] == [human: [name: "Luke"]]

**Comment by bbakerman** [comment]
I spoke to the lead of Sangria about there implementation and a couple of thigns stood out

First off the combined batching and @export as a featue.  

When they have say 3 operations. A, B,C they applied a dependency resolution process to decide the order.  

if they got a cyclic dep thet refuse the whole operation.

Otherwise thet worked out that it sould be B,C,A or A,C,B based on @exported variables

They streams the results out the other side. 

This allows for an order of say B,C,A and B can finish first as an execution result, then C and then A.

This allows you to send back B as results, and then C and then A.  In order words the story then comments then likes say??

**Comment by alexgenco** [comment]
Just wanted to bump this thread. Any reason not to revive this? We definitely have use cases for performing multiple operations in a single request.

**Comment by alexgenco** [comment]
> can you expand your use cases.

We don't have an immediate use case for _just_ executing multiple operations, except if it would make it easier for us to hack together an ad-hoc way to chain mutations with variables (something like the `@export` proposal). Also, we don't need to stream downstream responses, we would be ok with just waiting for all operations to finish to respond.

Bascially, we have designed our API in favor of small, composable mutations and queries, but clients are hesitant to eat the cost of 2-3 HTTP requests to get what they want.

Here's an example use case with actual mutations from our API:

```graphql
mutation Tokenize(input: TokenizeCreditCardInput!) {
  tokenizeCreditCard(input: $input) {
    paymentMethod {
      id @export(as: "singleUseId")
    }
  }
}

mutation Vault {
  vaultPaymentMethod(input: { paymentMethodId: $singleUseId }) {
    paymentMethod {
      id @export(as: "vaultedId")
    }
  }
}

mutation Charge {
  chargePaymentMethod(input: { paymentMethodId: $vaultedId }) {
    transaction {
      status
    }
  }
}
```

Basically, you tokenize a credit card (exchange raw details for a single-use token), then use that single-use token to permanently store the payment method in the merchant's vault, then use the vaulted payment method ID to charge the payment method.

We support a large number of payment method types (credit card, apple pay, google pay, etc.), and would prefer to not have to create a mutation for each of `tokenize`, `vault`, and `charge` for each payment method in order to limit the surface area of our API.

**Comment by andimarek** [comment]
closing this for now as multiple queries are not supported by the spec at the moment
